src/parser/state_machine/mod.rs

```diff
@@ -72,46 +72,73 @@ pub type StateResult = Result<(), ParsingTermination>;
 pub type ParseResult = Result<Never, ParsingTermination>;
 
 pub trait StateMachineActions {
-    fn emit_eof(&mut self, input: &[u8]) -> ActionResult;
-    fn emit_text(&mut self, input: &[u8]) -> ActionResult;
-    fn emit_current_token(&mut self, input: &[u8]) -> ActionResult;
-    fn emit_tag(&mut self, input: &[u8]) -> ActionResult;
-    fn emit_current_token_and_eof(&mut self, input: &[u8]) -> ActionResult;
-    fn emit_raw_without_token(&mut self, input: &[u8]) -> ActionResult;
-    fn emit_raw_without_token_and_eof(&mut self, input: &[u8]) -> ActionResult;
-
-    fn create_start_tag(&mut self, input: &[u8]);
-    fn create_end_tag(&mut self, input: &[u8]);
-    fn create_doctype(&mut self, input: &[u8]);
-    fn create_comment(&mut self, input: &[u8]);
-
-    fn start_token_part(&mut self, input: &[u8]);
-
-    fn mark_comment_text_end(&mut self, input: &[u8]);
-    fn shift_comment_text_end_by(&mut self, input: &[u8], offset: usize);
-
-    fn set_force_quirks(&mut self, input: &[u8]);
-    fn finish_doctype_name(&mut self, input: &[u8]);
-    fn finish_doctype_public_id(&mut self, input: &[u8]);
-    fn finish_doctype_system_id(&mut self, input: &[u8]);
-
-    fn finish_tag_name(&mut self, input: &[u8]) -> ActionResult;
-    fn update_tag_name_hash(&mut self, input: &[u8]);
-    fn mark_as_self_closing(&mut self, input: &[u8]);
-
-    fn start_attr(&mut self, input: &[u8]);
-    fn finish_attr_name(&mut self, input: &[u8]);
-    fn finish_attr_value(&mut self, input: &[u8]);
-    fn finish_attr(&mut self, input: &[u8]);
-
-    fn set_closing_quote_to_double(&mut self, input: &[u8]);
-    fn set_closing_quote_to_single(&mut self, input: &[u8]);
-
-    fn mark_tag_start(&mut self, input: &[u8]);
-    fn unmark_tag_start(&mut self, input: &[u8]);
-
-    fn enter_cdata(&mut self, input: &[u8]);
-    fn leave_cdata(&mut self, input: &[u8]);
+    type PassingState;
```

Comment:
Looking for a better name for this.

Reply:
I know it's overused as a term, but Context?

Reply:
Yeah, I think that's good: it is generic, but the nature of this type is also generic!

Reply:
yeah, was about to write that it's better to use `ParserContext`. Also, drop `passing` and `shared` prefixes. Just pass `ctx: ParserContext` around

Reply:
>why it needs to be a type parameter? is it because the state is generic over S.

It's because we need to reference it in the trait methods. (The ones in directly in the trait and the ones added by macros on the `StateMachine` "sub"-trait).

> Can we just make StateMachine trait generic then?

Not sure if I understand: the traits are generic over the context.